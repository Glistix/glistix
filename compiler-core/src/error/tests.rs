use super::*;
use insta::assert_snapshot;

#[test]
fn test_shell_program_not_found_error() {
    let cmds = vec![
        "erl", "erlc", "escript", "rebar3", "deno", "elixir", "node", "bun", "git",
    ];
    let oses = vec!["macos", "linux", "windows"];
    let distros = vec!["ubuntu", "unknown"];

    for cmd in &cmds {
        for os in &oses {
            match *os {
                "macos" | "windows" => {
                    let err = Error::ShellProgramNotFound {
                        program: cmd.to_string(),
                        os: os.to_string(),
                        distro: "unknown".to_string(),
                    }
                    .to_diagnostics();
                    assert_snapshot!(format!("{cmd}-{os}-unknown"), err[0].text);
                }
                "linux" => {
                    for distro in &distros {
                        let err = Error::ShellProgramNotFound {
                            program: cmd.to_string(),
                            os: os.to_string(),
                            distro: distro.to_string(),
                        }
                        .to_diagnostics();
                        assert_snapshot!(format!("{cmd}-{os}-{distro}"), err[0].text);
                    }
                }
                _ => (),
            }
        }
    }
}
