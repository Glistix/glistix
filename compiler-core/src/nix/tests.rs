use crate::ast::TypedModule;
use crate::config::PackageConfig;
use crate::type_::PRELUDE_MODULE_NAME;
use crate::{
    analyse::TargetSupport,
    build::{Origin, Target},
    nix::*,
    uid::UniqueIdGenerator,
    warning::{TypeWarningEmitter, WarningEmitter},
};
use camino::{Utf8Path, Utf8PathBuf};

mod assignments;
mod basic;
mod bit_arrays;
mod blocks;
mod bools;
mod case;
mod case_clause_guards;
mod consts;
mod custom_types;
mod externals;
mod functions;
mod lists;
mod modules;
mod numbers;
mod panic;
mod prelude;
mod records;
mod recursion;
mod results;
mod strings;
mod todo;
mod tuples;
mod type_alias;
mod use_;

pub static CURRENT_PACKAGE: &str = "thepackage";

#[macro_export]
macro_rules! assert_nix_with_multiple_imports {
    ($(($name:literal, $module_src:literal)),+; $src:literal) => {
        let compiled =
            $crate::nix::tests::compile_nix($src, vec![$((CURRENT_PACKAGE, $name, $module_src)),*]).expect("compilation failed");
        let mut output = String::from("----- SOURCE CODE\n");
        for (name, src) in [$(($name, $module_src)),*] {
            output.push_str(&format!("-- {name}.gleam\n{src}\n\n"));
        }
        output.push_str(&format!("-- main.gleam\n{}\n\n----- COMPILED NIX\n{compiled}", $src));
        insta::assert_snapshot!(insta::internals::AutoName, output, $src);
    };
}

#[macro_export]
macro_rules! assert_nix {
    (($dep_package:expr, $dep_name:expr, $dep_src:expr), $src:expr $(,)?) => {{
        let compiled =
            $crate::nix::tests::compile_nix($src, vec![($dep_package, $dep_name, $dep_src)])
                .expect("compilation failed");
        let output = format!(
            "----- SOURCE CODE\n{}\n\n----- COMPILED NIX\n{}",
            $src, compiled
        );
        insta::assert_snapshot!(insta::internals::AutoName, output, $src);
    }};

    (($dep_package:expr, $dep_name:expr, $dep_src:expr), $src:expr, $nix:expr $(,)?) => {{
        let compiled =
            $crate::nix::tests::compile_nix($src, Some(($dep_package, $dep_name, $dep_src)))
                .expect("compilation failed");
        let output = format!(
            "----- SOURCE CODE\n{}\n\n----- COMPILED NIX\n{}",
            $src, compiled
        );
        assert_eq!(($src, output), ($src, $nix.to_string()));
    }};

    ($src:expr $(,)?) => {{
        let output = $crate::nix::tests::compile_nix($src, vec![]).expect("compilation failed");
        insta::assert_snapshot!(insta::internals::AutoName, output, $src);
    }};

    ($src:expr, $js:expr $(,)?) => {{
        let output = $crate::nix::tests::compile_nix($src, vec![]).expect("compilation failed");
        assert_eq!(($src, output), ($src, $js.to_string()));
    }};
}

#[macro_export]
macro_rules! assert_nix_error {
    ($src:expr $(,)?) => {{
        let error = $crate::nix::tests::expect_nix_error($src, vec![]);
        let output = format!("----- SOURCE CODE\n{}\n\n----- ERROR\n{}", $src, error);
        insta::assert_snapshot!(insta::internals::AutoName, output, $src);
    }};
}

pub fn compile(src: &str, deps: Vec<(&str, &str, &str)>) -> TypedModule {
    let mut modules = im::HashMap::new();
    let ids = UniqueIdGenerator::new();
    // DUPE: preludeinsertion
    // TODO: Currently we do this here and also in the tests. It would be better
    // to have one place where we create all this required state for use in each
    // place.
    let _ = modules.insert(
        PRELUDE_MODULE_NAME.into(),
        crate::type_::build_prelude(&ids),
    );
    let mut direct_dependencies = std::collections::HashMap::from_iter(vec![]);

    deps.iter().for_each(|(dep_package, dep_name, dep_src)| {
        let mut dep_config = PackageConfig::default();
        dep_config.name = (*dep_package).into();
        let parsed = crate::parse::parse_module(
            Utf8PathBuf::from("test/path"),
            dep_src,
            &WarningEmitter::null(),
        )
        .expect("dep syntax error");
        let mut ast = parsed.module;
        ast.name = (*dep_name).into();
        let line_numbers = LineNumbers::new(dep_src);

        let dep = crate::analyse::ModuleAnalyzerConstructor::<()> {
            target: Target::Nix,
            ids: &ids,
            origin: Origin::Src,
            importable_modules: &modules,
            warnings: &TypeWarningEmitter::null(),
            direct_dependencies: &std::collections::HashMap::new(),
            target_support: TargetSupport::Enforced,
            package_config: &dep_config,
        }
        .infer_module(ast, line_numbers, "".into())
        .expect("should successfully infer");
        let _ = modules.insert((*dep_name).into(), dep.type_info);
        let _ = direct_dependencies.insert((*dep_package).into(), ());
    });

    let parsed =
        crate::parse::parse_module(Utf8PathBuf::from("test/path"), src, &WarningEmitter::null())
            .expect("syntax error");
    let mut ast = parsed.module;
    ast.name = "my/mod".into();
    let line_numbers = LineNumbers::new(src);
    let mut config = PackageConfig::default();
    config.name = "thepackage".into();

    crate::analyse::ModuleAnalyzerConstructor::<()> {
        target: Target::Nix,
        ids: &ids,
        origin: Origin::Src,
        importable_modules: &modules,
        warnings: &TypeWarningEmitter::null(),
        direct_dependencies: &direct_dependencies,
        target_support: TargetSupport::NotEnforced,
        package_config: &config,
    }
    .infer_module(ast, line_numbers, "".into())
    .expect("should successfully infer")
}

pub fn compile_nix(src: &str, deps: Vec<(&str, &str, &str)>) -> Result<String, crate::Error> {
    let ast = compile(src, deps);
    let line_numbers = LineNumbers::new(src);
    module(
        &ast,
        &line_numbers,
        Utf8Path::new(""),
        &"".into(),
        TargetSupport::Enforced,
    )
}

pub fn expect_nix_error(src: &str, deps: Vec<(&str, &str, &str)>) -> String {
    let error = compile_nix(src, deps).expect_err("should not compile");
    println!("er: {error:#?}");
    let better_error = match error {
        crate::Error::Nix {
            error: inner_error, ..
        } => crate::Error::Nix {
            src: src.into(),
            path: Utf8PathBuf::from("/src/nix/error.gleam"),
            error: inner_error,
        },
        _ => panic!("expected nix error, got {error:#?}"),
    };
    better_error.pretty_string()
}
