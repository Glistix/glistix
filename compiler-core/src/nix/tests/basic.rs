use crate::assert_nix;
use crate::nix::tests::CURRENT_PACKAGE;

#[test]
fn basic_test() {
    assert_nix!(
        r#"
pub type Amogus {
  AmogusA
  AmogusB(a: Int, b: Int)
  AmogusC(Int, Float)
}

pub const sumongus: Int = 5

pub fn sus(x: Int) -> Int {
    5 + 5
}

fn go(x, foo, boolthing, bmong) {
  let a = 1
  foo
  let a = 2
  let z = [fn(y: Int) { y }, foo(a, a, 50), fn(x: Int) { x }]
  let d = {
    "aaa"
  }
  let dd = {
    let amogus = 5.5e5
  }
  let intdiv = 5 / 5
  let floatdiv = 5.0 /. 5.0
  let remm = 5 % 5
  let g = 0.5
  let b = a
  foo(a, a, 50)
  let c = {
    let a = a
    [a]
  }
  foo
  let www = a == 5
  let wwww = {a + a} == {10 - 5 * 5}
  let first_list = [1 < 4, 3 >= 5]
  let list_thing = [www, ..first_list]
  let pipes =
    a
    |> foo(_, a, 50)
    |> fn(x) { x }
  let pipes2 = a |> fn(i: Int) { i }
  panic as "amongus"
  todo as "amongus"
  panic
  todo
  // let mongus: Amogus = bmong
  // let mongus = bmong.a
  // let mongus = Amogus(..bmong, a: 6)
  let y = -500 + 10 - {-a} - {-5}
  let yy = [ -a, -5, -y ]
  let f = -5.5 -. 5.2e-5
  let f = 5.5 -. 5.2e5
  // TODO: fix string escaping
  let ss = "a" <> "b" <> "c d"
  let ff = [ -5.2e-5, -5.5 ]
  let z = !boolthing
  let tupdatup = #(1, 2, "a", -5.5, #(1, 2, 3))
  let tata = tupdatup.1 + tupdatup.0 + 20
  let mabool = True
  let mnotbool = False
  let amagus_a = AmogusA
  let amagus_b = AmogusB(10, b: 15)
  let amagus_c = AmogusC(10, 5.5)
  let simplefunc = fn() { 5 }
  let gg = simplefunc() + 10
  let less_simple_func = fn(x) { fn() { x } }
  let gg = less_simple_func(5)() + 10
  let mnull = Nil
  let sus = "hey! I'm still sus :("
  // make sure arguments are counted in initial state
  let x = tata
  x
}

fn mongus(a, b, c) {
  let x = fn() { 5 + 5 }
  let y = x()
  #(1, 2, 3, fn(x: Int) { x + 1 }, { 5 6 10 {10 + 5} })
}
"#,
    )
}

#[test]
fn basic_case_test() {
    assert_nix!(
        r#"
fn condman(a, b) {
  let p1 = case a > b {
    True -> 5
    False -> 6
  }

  let p2 = case a > b {
    True -> 5
    _ -> 6
  }

  let p3 = case a > b {
    False -> 6
    True -> 5
  }

  let p4 = case a > b {
    False -> 6
    _ -> 5
  }

  let v1 = True
  let v2 = False
  let vv1 = case v1 { _ -> 10 }
  let vv2 = case v2 { True -> 50 _ -> 10 }
  let vv3 = case True { True -> 50 _ -> 10 }

  #(p1, p2, p3, p4, vv1, vv2, vv3)
}

pub fn fact(n: Int) -> #(Nil, Int) {
  let res = case n < 0 {
    True -> panic as "Don't."
    False -> Nil
  }
  #(res, case n {
    0 -> 1
    _ -> n * fact(n - 1).1
  })
}
"#
    )
}

#[test]
pub fn basic_diverse_case_test() {
    assert_nix!(
        r#"
pub type SimpleEnum {
  SA
  SB
  SC
  SD
}

fn simple_test(x: SimpleEnum) {
  case x {
    SC -> Ok(Ok(100))
    _ -> Error(Error(100))
  }

  case x {
    SA -> Ok(5)
    SB -> Error(10)
    SC -> Ok(10)
    SD -> Ok(10)
  }
}

pub fn simple_test2(x: SimpleEnum, y: Int, z: Float, w: String, p: Nil) {
  let x1 = case x {
    SC -> "is SC"
    SA -> "is SA"
    _ -> "not SC or SA"
  }
  let y1 = case y {
    0 -> "is 0"
    -1 -> "is -1"
    100 -> "is 100"
    _ -> "unknown int"
  }
  let z1 = case z {
    0.0 -> "is 0.0"
    53.53 -> "is 53.53"
    1.0e2 -> "is 1.0e2"
    _ -> "unknown float"
  }
  let w1 = case w {
    "a" -> "is a"
    "b" -> "is b"
    "c" -> "is c"
    _ -> "not a, b, c"
  }
  let p1 = case p {
    Nil -> "is nil"
    _ -> "not nil (impossible)"
  }
  #(x1, y1, z1, w1, p1)
}
"#
    )
}

#[test]
pub fn basic_assigning_case_test() {
    assert_nix!(
        r#"
pub type MyRecord {
  Simple
  TupleLike(Int, Float)
  RecordLike(a: Int, inherit: Float)
  Mixed(a: Int, b: Float, Float, Float)
}

pub fn please_match(x) {
  case x {
    Simple as s -> {s #(-100, -100.0)}
    TupleLike(i, f) as t -> { t #(i, f) }
    RecordLike(a: i, inherit: f) -> { #(i, f) }
    Mixed(f1, f2, a: i1, b: f3) -> #(i1, f1 +. f2 +. f3)
    t -> { t #(555, 555.5) }
  }
}
"#
    )
}

#[test]
pub fn basic_case_guard_test() {
    assert_nix!(
        (CURRENT_PACKAGE, "mymod", r#"pub const inherit = False"#),
        r#"
import mymod

type BigBad {
  BigBad(a: Result(Int, Result(Int, Int)), b: Int)
}

pub const inherit = 5

pub const bad = [Ok(5), Ok(6)]

const bigbad = [BigBad(Error(Ok(5)), 10), BigBad(Error(Ok(5)), 10)]

pub type Bad {
  Bad(inherit: Bool)
}

pub fn guard(x) {
  let b = Bad(inherit: True)
  case x {
    Nil if mymod.inherit -> 1
    Nil if b.inherit -> 2
    Nil if Bad(inherit: True) == Bad(inherit: False) -> 3
    Nil if BigBad(Error(Ok(5)), 10) == BigBad(Error(Ok(5)), 10) -> 4
    Nil if 5 > 6 -> 5
    Nil if [Ok(5), Ok(6)] == [Ok(7)] -> 6
    _ -> 7
  }
}
"#
    )
}
