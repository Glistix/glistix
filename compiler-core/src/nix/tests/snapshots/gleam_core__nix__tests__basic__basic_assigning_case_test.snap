---
source: compiler-core/src/nix/tests/basic.rs
expression: "\npub type MyRecord {\n  Simple\n  TupleLike(Int, Float)\n  RecordLike(a: Int, inherit: Float)\n  Mixed(a: Int, b: Float, Float, Float)\n}\n\npub fn please_match(x) {\n  case x {\n    Simple as s -> {s #(-100, -100.0)}\n    TupleLike(i, f) as t -> { t #(i, f) }\n    RecordLike(a: i, inherit: f) -> { #(i, f) }\n    Mixed(f1, f2, a: i1, b: f3) -> #(i1, f1 +. f2 +. f3)\n    t -> { t #(555, 555.5) }\n  }\n}\n"
---
let
  Simple = { __gleam_tag' = "Simple"; };
  
  TupleLike = x0: x1: { __gleam_tag' = "TupleLike"; _0 = x0; _1 = x1; };
  
  RecordLike =
    a: inherit': { __gleam_tag' = "RecordLike"; inherit a inherit'; };
  
  Mixed =
    a: b: x2: x3: { __gleam_tag' = "Mixed"; inherit a b; _2 = x2; _3 = x3; };
  
  please_match =
    x:
      if x.__gleam_tag' == "Simple"
      then let s = x; in let _''0 = s; in { _0 = -100; _1 = -100.0; }
      else if x.__gleam_tag' == "TupleLike"
      then
        let
          t = x;
          i = x._0;
          f = x._1;
        in let _''0 = t; in { _0 = i; _1 = f; }
      else if x.__gleam_tag' == "RecordLike"
      then let i = x.a; f = x.inherit'; in { _0 = i; _1 = f; }
      else if x.__gleam_tag' == "Mixed"
      then
        let
          i1 = x.a;
          f3 = x.b;
          f1 = x._2;
          f2 = x._3;
        in { _0 = i1; _1 = (f1 + f2) + f3; }
      else let t = x; in let _''0 = t; in { _0 = 555; _1 = 555.5; };
in { inherit Simple TupleLike RecordLike Mixed please_match; }
